Export as RecordBatch async iterator

rdettai opened the pull request:
For performance and typing reasons, we want all exports that occur in the python bindings to go through Arrow. The new signature of `export` should be:
```
async def export(expression) -> AsyncIterable[TableSlice]:
```
Where:
- `TableSlice` is just an indirection for `pyarrow.RecordBatch` (to avoid having an API level dependency on PyArrow)
- The output is an async iterable because exports can be both historical or continuous

We are also getting rid of the current `export_row` workaround (using json export) to have a unified export path.

## Reviewer's instruction

### Tests in docker

Tests that require the VAST binary run in docker. I wanted to provide a bit of scaffolding to make the dev experience a bit nicer. Unfortunately, you cannot run arbitrary bash in Poetry scripts, so I added a separate script:
```
./docker-poetry-run.sh pytest -v
```
The name is verbose but at least it is explicit.

### Maintaining TheHive integration

TheHive integration already had some stability issues. We are going to assume that having one successful run in CI is good enough to validate that the integration is compatible with the new export path.

rdettai commented:
When converting the record batch associated with the alert following alert from `vast/integration/data/suricata/eve.json`:
``` 
{"timestamp":"2011-08-12T14:52:57.716360+0200","flow_id":1031464864740687,"pcap_cnt":83,"event_type":"alert","src_ip":"147.32.84.165","src_port":1181,"dest_ip":"78.40.125.4","dest_port":6667,"proto":"TCP","alert":{"action":"allowed","gid":1,"signature_id":2017318,"rev":4,"signature":"ET CURRENT_EVENTS SUSPICIOUS IRC - PRIVMSG *.(exe|tar|tgz|zip)  download command","category":"Potentially Bad Traffic","severity":2},"flow":{"pkts_toserver":27,"pkts_toclient":35,"bytes_toserver":2302,"bytes_toclient":4520,"start":"2011-08-12T14:47:24.357711+0200"},"payload":"UFJJVk1TRyAjemFyYXNhNDggOiBzbXNzLmV4ZSAoMzY4KQ0K","payload_printable":"PRIVMSG #zarasa48 : smss.exe (368)\r\n","stream":0,"packet":"AB5J2xnDCAAntbcZCABFAABMGV5AAIAGLlyTIFSlTih9BASdGgvw0QvAxUWHdVAY+rCL4gAAUFJJVk1TRyAjemFyYXNhNDggOiBzbXNzLmV4ZSAoMzY4KQ0K","packet_info":{"linktype":1}}
```

I get the following exception when calling `to_pylist()`:
```
self = <[AttributeError("'NoneType' object has no attribute 'as_py'") raised in repr()] AddressScalar object at 0x7f36b6946930>

   def as_py(self) -> ip.IPv4Address | ip.IPv6Address:
>       return unpack_ip(self.value.as_py())
E       AttributeError: 'NoneType' object has no attribute 'as_py'
```

Looks like the cause is that the `suricata.alert` type contains a nested record:
```
    alert.source: record {
      ip: addr,
      port: port
    },
```

That is null in the provided JSON sample. 

@mavam I know that we already crossed similar issues with nested records, but I can't find any trace of it anymore. Do you have pointers to what it was exactly?

mavam commented:
It'd be good to see whether we're using the API wrong or whether the idiom is that `self.value` must be checked *within* `as_py` to be `Null`.

If that's the case, we may have to adapt our other extension type definitions.

dominiklohmann commented:
> It'd be good to see whether we're using the API wrong or whether the idiom is that `self.value` must be checked _within_ `as_py` to be `Null`.

That would mirror my understanding of the C++ APIs, where we have to do the same before accessing a value in an array.

rdettai commented:
@mavam this is now rebased on top of the extension types fixes https://github.com/tenzir/vast/pull/2816

rdettai commented:
Unfortunately this is still failing on the continuous query use case. Further investigation is required.

rdettai commented:
@tobim everything works as expected apart from TheHive integration. I'll take a look at that on Monday. Until then, if you wish, you can already go through the rest of the PR (everything except `docker/thehive/app/app.py` and `docker/compose/thehive.app.yaml`). Thanks!

tobim commented on python/tests/test_vast.py:
```diff
@@ -1,7 +1,121 @@
+    logger.debug(f"vast flush stderr:\n{flush_err.decode()}")
+
+
+def integration_data(path):
+    return f"../../vast/integration/data/{path}"
```
```suggestion
def integration_data(path):
    dir_path = os.path.dirname(os.path.realpath(__file__))
    return f"{dir_path}/../../vast/integration/data/{path}"
```

tobim reviewed (commented):
Just a bit of feedback. 
I'd really like to be able to run the integration tests without docker.
Also, I'm not sure about the `TableSlice` wrapper.

rdettai commented on python/vast/vast/vast.py:
```diff
@@ -37,78 +102,45 @@ async def start():
+                logger.debug(f"[export-{id}] {line.decode().strip()}")
+            logger.debug(f"[export-{id}] Stop logging")
+
+        # TODO: don't use default thread pool here as it won't scale
```
Good question! the default executor is [`concurrent.futures.ThreadPoolExecutor`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor). This has a default max thread (proc on the machine x5). It will not grow indefinitely and we might quickly be limited if we run multiple exports in parallel. Using a thread here really defeats the purpose of asyncio 😢. There is a [workaround](https://github.com/cloudfuse-io/lambdatization/blob/eb29b12ce0363415bfafca2d09ec605d705dd814/docker/clickhouse/lambda-handler.py#L18-L63) that uses selectors and would make it possible to do all the logging on a single thread, but I really don't want to over-engineer right away + make this PR to big

tobim replied:
TIL. Can you change the comment to something like
```suggestion
        # TODO: don't use default thread pool here as this method 
        # will block one thread per running export, risking exhaustion of 
        # the pool if we run many in parallel.
```

rdettai commented on python/vast/vast/vast.py:
```diff
@@ -37,78 +102,45 @@ async def start():
+        t = asyncio.create_task(asyncio.to_thread(log))
+
+        try:
+            # VAST concatenates IPC streams for different types so we need to
+            # spawn multiple stream readers
```
Flight RPC or anything that encapsulates the streams better. Resolving this as this is outside the scope of this PR. Let's discuss it separately.

rdettai commented on python/vast/vast/vast.py:
```diff
@@ -37,78 +102,45 @@ async def start():
+            if isinstance(e, pa.ArrowInvalid):
+                logger.debug("completed processing stream of record batches")
+            else:
+                proc.kill()
+                await asyncio.wait_for(t, 3)
```
> Assuming the wait gets a timeout in seconds.. Why 3?
3 because I wanted it to be more than 2 but less than 4 😄 

Joke aside:
> Is there a reason to KILL instead of TERM here?
We are already in an unexpected error case, so the there is no point in being graceful in the exit. We just want to get out of here 
> Shouldn't log be guaranteed to exit, so we could just do a regular wait?
the wrapping with a timeout comes from a general distrust in the current state as we are in an unexpected error. We know we failed, there is no point in taking the risk of hanging on top of it. I think log is guaranteed to exit, but there is nothing worth than debugging something that is hanging so I prefer to be safe and exit. I don't know what time to choose here. 3 seems more than enough for the logger to wrap up while failing relatively fast if things are hanging. I should provide a small explanation of this logic.

rdettai replied:
```suggestion
                # Use SIGKILL as we are already in an unexpected error state, so
                # there is no point in trying a graceful exit.
                proc.kill()
                # We set a timeout to make sure that we don't hang while tearing
                # down the logging task. The duration of 3s is arbitrary and is
                # expected to be more than enough for the logger task to
                # complete.
                await asyncio.wait_for(t, 3)
                raise e
```

rdettai commented on python/vast/vast/vast.py:
```diff
@@ -22,12 +21,78 @@ class ExportMode(Enum):
     UNIFIED = auto()
 
 
+class TableSlice(ABC):
+    """The VAST abstraction wrapping Arrow record batches"""
+
+
+class PyArrowTableSlice(TableSlice):
+    """A TableSlice internally represented by a PyArrow RecordBatch"""
+
+    def __init__(self, batch: pa.RecordBatch):
+        self._batch = batch
```
To summarize or discussion
- we don't want our API to be explicitly bound to `pyarrow`
  - we might want to swap the bindings we use to interact with arrow without impacting the general API
  - in general, and especially when they interact with the data row wise, users don't really need to know that we are using arrow.

rdettai commented on docker/dev/README.md:
```diff
@@ -2,18 +2,22 @@
 
-A helper to build Poetry based Python integrations between VAST and third party
-tools.
+A helper image to build Python integrations on top of VAST without having to
+install the Poetry dependencies locally.
```
This is used by the current way we package apps (like TheHive) with Docker. But I think that now that we cleaned up the fabric code, I could actually bring that code back to the `python` directory.

rdettai replied:
I move thehive python code to the `python/` dir and it simplified this part greatly. This Dockerfile is still used for tests and for building/running apps in compose, but it has been simplified quite a bit

rdettai commented:
TODO: adapt thehive to use to_json_rows

rdettai commented:
@tobim the main things that changed since your last review ([42d9745](https://github.com/tenzir/vast/pull/2791/commits/42d9745a384d88ee5938500b02cf6dd7d12b6716) to [22ee91f](https://github.com/tenzir/vast/pull/2791/commits/22ee91f1d853ed2e273c4977a7093cf9d44e610d)):

- I added a `to_json_rows` utility function to enable integrations to use dumbed down row by row data. The hard part was actually to write the test that contained all Arrow types (`test_arrow_dict_to_json_dict_native_types`).
- in TheHive app I removed some unnecessary caching logic and did the required change to use the new `to_json_rows` 
- I refactored the TheHive app to be part of the `python/` dir. They were initially separated because VAST python contained a some legacy code that made it confusing to add apps on top, but now that it was cleaned up it makes sense to centralize our python code into a single location

tobim commented on python/vast/vast/convert.py:
```diff
@@ -1,7 +1,11 @@
+from decimal import Decimal
+from typing import Any, AsyncIterable
 
+import numpy as np
```
Do we have to add `oldest-supported-numpy` to `pyproject.toml`?

rdettai replied:
No, but they are a dependency of pandas that do have:
```
  "numpy>=1.20.3; python_version<'3.10'",
  "numpy>=1.21.0; python_version>='3.10'",
  "numpy>=1.23.2; python_version>='3.11'",
```

What is your concern exactly?

tobim replied:
We use `numpy` directly, so we should list it as a dep, no?

rdettai replied:
ah yeah, I think you are right, that would be cleaner. Even though we are just using it to validate the output of a dep that uses it, I guess it is best practice to list all deps that are explicitly imported

tobim reviewed (approved):
This looks great. Only one question about numpy.

rdettai commented:
Part of https://github.com/tenzir/public-roadmap/issues/3
